### PR TITLE
bpo-40397: Refactor typing._GenericAlias

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue40397

### CODE REVIEW

**Comment by terryjreedy** [comment]
From a cursory reading, the main substantive change is to remove the 'special' attribute from the Generic class and move the 'special' code to the Special class and its subclass.  The net result is no change in the number of lines.  I am pretty sure I would prefer carefully reading the new version.  I would prefer being able to ignore and then focus on the Special stuff, especially if there were an explanatory docstring at the top.  How good is the test coverage?  (Don't have my Python machine.)

**Comment by serhiy-storchaka** [comment]
Yes, it is the main change - splitting the `_GenericAlias` class on two classes, `_SpecialGenericAlias` and new `_GenericAlias` which are split on more specialized subclasses. All special cases in common methods are3 replaced with overriding methods in subclasses. In future the new `_GenericAlias` will be replaced with `GenericAlias`, and `_SpecialGenericAlias` will be more simplified (for example it does not need `__args__` and `__parameters__`). Further special generic aliases in `typing` like `List`, `AbstractSet` and `Callable` may be replaced with original types (builtin `list`, `collections.abc.Set`, `collections.abc.Callable`). `_UnionGenericAlias` may be made a builtin (to support `type | type`).

Most thing are tested, and this PR does not change the behavior. Thing which are not tested I consider undefined.

File: Lib/typing.py [hunk 1]
```diff
@@ -801,14 +730,70 @@ def __reduce__(self):
...
+        if not isinstance(params, tuple):
+            params = (params,)
+        msg = "Parameters to generic types must be types."
+        params = tuple(_type_check(p, msg) for p in params)
+        _check_generic(self, params)
+        assert self.__args__ == self.__parameters__
+        return self.copy_with(params)
+
+    def copy_with(self, params):
+        return _GenericAlias(self.__origin__, params,
```
Comment (gvanrossum) [inline Lib/typing.py:760]: Why not `_SpecialGenericAlias(...)`? (If there's a reason, add a comment.)

**Review by gvanrossum — COMMENTED** [review]
I don't feel comfortable reviewing, hopefully Ivan can. (Though I approve of the spirit of the change!)

File: Lib/typing.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/typing.py:760]: `_SpecialGenericAlias` is for non-parametrized aliases, like `List`. `_GenericAlias` is for parametrized aliases, like `List[int]`. By design, when you you subscribe `_SpecialGenericAlias` you get `_GenericAlias`. `_GenericAlias` is like `GenericAlias` with some additions (for example it checks and converts arguments). `_SpecialGenericAlias` is a producer of `_GenericAlias` (other producer is `_SpecialForm` -- class for `Union`, `ClassVar`, etc).

Comment (gvanrossum) [reply Lib/typing.py:760]: Ah, got it. `List.copy_with(X)` returns `List[X]`.

**Comment by serhiy-storchaka** [comment]
I think this PR is ready for review. It is a first step which should not change any behavior. In the following steps I am going to change some unspecified and untested behavior.

Please make a review.

**Review by gvanrossum — APPROVED** [review]
LGTM. Glad to see this going a bit more OO...

**Pull request merged**